File: application/contracts/response/error_message.py
from typing import TypedDict, Literal, Union, Optional

ErrorType = Union[Literal["invalid credentials"],
                  Literal["invalid token"],
                  Literal["internal error"],
                  Literal["permission denied"],
                  Literal["nickname already used"],
                  Literal["login already used"],
                  Literal["banned"],
                  Literal["not authorized"]]


class ErrorMessageResult(TypedDict):
    message: str
    error_type: ErrorType


class ErrorMessage(TypedDict):
    type: Literal["error"]
    result: ErrorMessageResult


class ErrorMessageWithId(ErrorMessage):
    id: Optional[int]

File: frontend/src/contracts/messages/response/error-message.ts
import IBaseResponseMessage from "../base-response-message";

type ErrorType =
    | "invalid credentials"
    | "invalid token"
    | "internal error"
    | "permission denied"
    | "nickname already used"
    | "login already used"
    | "not authorized"
    | "banned";

export default interface IErrorMessage extends IBaseResponseMessage {
    type: "error";
    result: {
        message: string;
        error_type: ErrorType;
    }
}